nav2_map_server/include/nav2_map_server/map_saver.hpp

```diff
@@ -16,56 +17,77 @@
 #define NAV2_MAP_SERVER__MAP_SAVER_HPP_
 
 #include <string>
-#include "map_mode.hpp"
-#include "nav_msgs/srv/get_map.hpp"
+#include <memory>
+
 #include "rclcpp/rclcpp.hpp"
+#include "nav2_util/lifecycle_node.hpp"
+#include "nav2_msgs/srv/save_map.hpp"
+
+#include "mapio.hpp"
 
 namespace nav2_map_server
 {
 
 /**
  * @class nav2_map_server::MapSaver
- * @brief A class that writes map to a file from occpancy grid which is
- * subscribed from "/map" topic.
+ * @brief A class that provides map saving methods and services
  */
-class MapSaver : public rclcpp::Node
+class MapSaver : public nav2_util::LifecycleNode
 {
 public:
   /**
-   * @brief Constructor for the MapSaver
-   * @param options NodeOptions for the MapSaver
+   * @brief Constructor for the nav2_map_server::MapSaver
    */
-  explicit MapSaver(const rclcpp::NodeOptions & options);
+  MapSaver();
 
   /**
-   * @brief A Map callback function calls try_write_map_to_file method to
-   * write map data to a file.
-   * @param map Occupancy Grid message data
+   * @brief Destructor for the nav2_map_server::MapServer
    */
-  void mapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr map);
+  ~MapSaver();
 
   /**
-   * @brief Returns the promise as shared future
-   * @return a shared future copy of save_next_map_promise
+   * @brief Read a message from incoming map topic and save map to a file
+   * @param map_topic Incoming map topic name
+   * NOTE: map_topic could be updated during function execution.
+   * @param save_parameters Map saving parameters.
+   * NOTE: save_parameters could be updated during function execution.
+   * @return true of false
    */
-  std::shared_future<void> map_saved_future() {return save_next_map_promise.get_future().share();}
+  bool saveMapTopicToFile(std::string & map_topic, SaveParameters & save_parameters);
```

Comment:
please make parameters that this function doesn't implicitly modify as `const`

Reply:
`saveMapTopicToFile()` changes both `map_topic` and `save_parameters` parameters by its logic: these parameters are checked for correctness inside this method and being updated if necessary. Do they really need to be a constant?

Reply:
Yes, its unexpected behavior to have a function change your parameters. Imagine if you have multiple things being saved, then the inputs will have their state messed up between them. 

It looks like you change values from error checking. I think its reasonable to incur a copy here and modify parameters on the copies (e.g. either don't pass by reference or copy in the function and modify the copied versions).

Reply:
Ok, made API changes for `saveMapTopicToFile()` - https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-e22d88faf6e8ac7a5338c8c6cbc2f7a4R55 and `saveMapToFile()` - https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-b3fa8804c8edd37f085c3ee28e9a092bR99